gourmet/timer.py

```diff
@@ -1,99 +1,132 @@
 
-    def set_time (self, s):
-        s = int(s)
-        if s < 0: s = 0
-        #self.hoursSpin.set_text(self.pad_n(s / 3600))
-        self.hoursSpin.set_value(s / 3600)
-        s = s % 3600
-        #self.minutesSpin.set_text(self.pad_n(s / 60))
-        self.minutesSpin.set_value(s / 60)
-        #self.secondsSpin.set_text(self.pad_n(s % 60))
-        self.secondsSpin.set_value(s % 60)
-
-    def val_changed_cb (self, widg):
+    def val_changed_cb(self, widg: Gtk.SpinButton) -> None:
+        """On input callback to set the values to be always two digits"""
         if not widg.val_change_is_changing_entry:
             widg.val_change_is_changing_entry = True
-            widg.set_text(self.pad_n(int(widg.get_value())))
+            widg.set_text(widg.get_text().zfill(2))
             widg.val_change_is_changing_entry = False
 
-    def pad_n (self, int):
-        s = str(int)
-        if len(s)==1: return '0'+s
-        else: return s
+    def tick(self) -> bool:
+        """Run the timer
 
-    def get_time (self):
-        return self.hoursSpin.get_value()*3600 + self.minutesSpin.get_value()*60 + self.secondsSpin.get_value()
+        This is done in two steps: first we compute the time elapsed between two
+        ticks of the loop, and then we subtract it from the timer time, which is
+        then updated.
+
+        Returns a bool to notify the event-loop on whether the timer is done.
+        """
+        if self.is_running:
+            if self.first_iteration:
+                self.first_iteration = False
+                self.previous_iter_time = time.time()
+
+            now = time.time()
+            elapsed = now - self.previous_iter_time
+            self.previous_iter_time = now
 
-    # Methods to run the timer...
-    def tick (self):
-        if self.running:
-            elapsed = time.time() - self.running
-            t = self.start_time - elapsed
-            #t = self.get_time() - 1
-            self.set_time(t)
-            if t<=0:
+            t = self.get_time()
+            remaining = t - elapsed
+            self.set_time(remaining)
+
+            if remaining <= 0:
                 self.finish_timer()
                 return False
             else:
                 return True
         else:
             return False
 
-    def start_cb (self,*args):
-        if not self.running and self.get_time():
-            self.running = time.time()
-            self.orig_time = self.start_time = self.get_time()
-            GObject.timeout_add(1000,self.tick)
-
-    def pause_cb (self, *args):
-        if self.running:
-            self.running = False
-        else:
-            self.running = time.time()
-            self.start_time = self.get_time()
-        if self.running: GObject.timeout_add(1000,self.tick)
-
-    def reset_cb (self, *args):
-        self.running = False
+    def start_cb(self, *args) -> None:
+        if not self.is_running:
+            self.previous_iter_time = 0
```

Comment:
It seems to me that `previous_iter_time` is exactly `0` when `first_iteration` is `False`. Maybe we want to set `previous_iter_time` to `None` when we mean `first_iteration`?

Reply:
Cool! Thanks for the idea